src/mlpack/core/optimizers/cne/cne_impl.hpp

```diff
@@ -0,0 +1,202 @@
+    mutationSize(mutationSize),
+    selectPercent(selectPercent),
+    finalValue(finalValue),
+    fitnessHist(fitnessHist)
+{ /* Nothing to do here. */ }
+
+//! Optimize the function
+template<typename DecomposableFunctionType>
+double CNE::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // Get the number of functions to iterate
+  const size_t numFun = function.NumFunctions();
+
+  // Set the population size and fill random values [0,1]
+  population = arma::randu(populationSize, iterate.n_rows);
+
+  // initialize helper variables
+  fitnessValues.set_size(populationSize);
+  double fitness = 0;
+  double lastBestFitness;
+
+  Log::Info << "CNE initialized successfully. Optimization started"
+      << std::endl;
+
+  // Iterate till max number of generations
+  for (size_t gen = 1; gen <= maxGenerations; gen++)
+  {
+    // calculate fitness values of all candidates
+    for (size_t i = 0; i < populationSize; i++)
+    {
+       // select a candidate and insert the parameters in the function
+       iterate = population.row(i).t();
+
+       // find the fitness
+       for (size_t j = 0; j < numFun; j++)
+         fitness += function.Evaluate(iterate, j);
+
+       // Save fitness values
+       fitnessValues[i] = fitness;
+       fitness = 0;
+    }
+
+      Log::Info << "Generation number: " << gen << " best fitness = "
+          << fitnessValues.min() << std::endl;
+
+      // create the next generation of species
+      Reproduce();
+
+      // check for termination
+      if (finalValue != DBL_MIN && finalValue >= fitnessValues.min())
+      {
+          Log::Info << "Terminating. Given fitness criteria " << finalValue
+              << " > " << fitnessValues.min() << std::endl;
+          break;
+      }
+
+      // check for termination
+      if (fitnessHist != DBL_MIN && gen != 1 &&
+        (lastBestFitness - fitnessValues.min()) < fitnessHist)
+      {
+          Log::Info << "Terminating. Fitness History change "
+              << (lastBestFitness - fitnessValues.min())
+          << " < " << fitnessHist << std::endl;
+          break;
+      }
+
+      // Store the best fitness of this generation before update
+      lastBestFitness = fitnessValues.min();
+  }
+
+  // Set the best candidate into the network
+  iterate = population.submat(index[0], 0, index[0], iterate.n_rows - 1).t();
+
+  // find the best fitness
+  for (size_t j = 0; j < numFun; j++)
+      fitness += function.Evaluate(iterate, j);
+
+  return fitness;
+}
+
+//! Reproduce candidates to create the next generation
+void CNE::Reproduce()
+{
+  // Sort fitness value. The smaller the better
+  index = arma::sort_index(fitnessValues);
+
+  // Find the number of elite percentage
+  size_t numElite = floor(selectPercent * populationSize);
+
+  // Making sure we have even number of candidates to remove and create
+  if ((populationSize - numElite) % 2 != 0) numElite++;
+
+  for (size_t i = numElite; i < populationSize-1; i++)
```

Comment:
Pedantic style issue; missing space between ``populationSize`` and the operation.